rust/eos-rs/src/api/dungeon_mode/dungeon_struct.rs

```diff
@@ -0,0 +1,990 @@
+    ///
+    /// # Safety
+    /// This is unsafe, since it updates the global dungeon struct (essentially a `static mut`).
+    pub unsafe fn z_init(&mut self) {
+        ffi::DungeonZInit()
+    }
+
+    /// Frees the dungeons struct pointer to by the master dungeon pointer,
+    /// and nullifies the pointer.
+    ///
+    /// # Safety
+    /// This is unsafe, since it completely invalidates borrows of the old global dungeon struct
+    /// and then invalidates the global dungeon pointer.
+    pub unsafe fn free(self) {
+        ffi::DungeonFree()
+    }
+
+    /// Returns a reference to the inner dungeon struct.
+    pub fn inner(&self) -> &Dungeon<&'a mut ffi::dungeon> {
+        &self.1
+    }
+
+    /// Returns a mutable reference to the inner dungeon struct.
+    pub fn inner_mut(&mut self) -> &mut Dungeon<&'a mut ffi::dungeon> {
+        &mut self.1
+    }
+
+    /// Seems to initialize the dungeon struct from specified dungeon data.
+    ///
+    /// The signature will be updated once we know more about this function.
+    pub fn initialize_dungeon(
+        &mut self,
+        dungeon_data: &mut ffi::undefined,
+        dungeon: &mut ffi::dungeon,
+    ) -> i32 {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::InitializeDungeon(dungeon_data as *mut _, dungeon as *mut _) }
+    }
+
+    /// Returns an abstraction over the game's builtin dungeon generator. This
+    /// struct implements [`dungeon_generator::DungeonFloorGeneration`], which is the
+    /// recommended way to work with it, see the documentation of [`dungeon_generator`].
+    ///
+    /// # Note
+    /// Note that the builtin dungeon generator works on the global dungeon struct directly.
+    pub fn get_builtin_dungeon_generator(
+        &'a mut self,
+    ) -> dungeon_generator::game_builtin::GlobalDungeonStructureGenerator<'a> {
+        dungeon_generator::game_builtin::GlobalDungeonStructureGenerator(self.0.clone(), self)
+    }
+
+    /// Generates a dungeon floor.
+    ///
+    /// If not changed by a patch, this function will use the game's default built in generator
+    /// and generate the floor based on the current global configuration for the floor.
+    ///
+    /// For more granular control over the floor generation, you can get an abstraction over
+    /// the builtin generator with [`Self::get_builtin_dungeon_generator`].
+    ///
+    pub fn generate_floor(&'a mut self) {
+        self.get_builtin_dungeon_generator()
+            .generate_floor_internal();
+    }
+
+    /// Gets the floor type. Returns None if the global dungeon struct contains invalid data.
+    pub fn get_floor_type(&self) -> Option<FloorType> {
+        // SAFETY:We hold a valid mutable reference to the global dungeon struct.
+        unsafe { ffi::GetFloorType() }.try_into().ok()
+    }
+
+    /// Checks if the current fixed floor is the "substitute room" (Fixed Room ID 0x6E).
+    pub fn is_substitute_room(&self) -> bool {
```

Comment:
Putting a definition of "substitute room" would be a good idea (The only reference I find is in @UsernameFodder work on `pmdsky-debug`, with no more indication on what it is. Which have the same definition). Is that the room that is used when generation failed ?

Reply:
It's literally a room with substitutes in it. I feel as though specifying "Fixed Room ID 0x6E" is about as precise as you could get though.